soccer/vision/camera/Camera.cpp

```diff
@@ -0,0 +1,389 @@
+    }
+
+    // We have some balls, so choose which updater to use
+    if (*use_MHKF) {
+        updateBallsMHKF(calcTime, ballList, previousWorldBall);
+    } else {
+        updateBallsAKF(calcTime, ballList, previousWorldBall);
+    }
+}
+
+void Camera::updateBallsMHKF(RJ::Time calcTime,
+                             const std::vector<CameraBall> ballList,
+                             const WorldBall& previousWorldBall) {
+    // If we have no existing filters, create a new one from average of everything
+    // Easier than trying to figure out which ones are more than X meters away from each other
+    // Only delays the filter collection by a camera frame or two
+    if (kalmanBallList.size() == 0) {
+        std::list<CameraBall> ballList2;
+        std::copy(ballList.begin(), ballList.end(), std::back_inserter(ballList2));
+        
+        CameraBall avgBall = CameraBall::CombineBalls(ballList2);
+        kalmanBallList.emplace_back(cameraID, calcTime, avgBall, previousWorldBall);
+
+        return;
+    }
+
+    // TODO: Try merging some of the kalman filters together
+
+    // Create list of bools corresponding to whether we have used this ball
+    // as measurement yet
+    std::vector<bool> usedCameraBall(ballList.size(), false);
+
+    // Which camera balls to apply to which kalman ball
+    std::vector<std::list<CameraBall>> appliedBallsList(kalmanBallList.size());
+
+    // Figure out which measurements go with which kalman balls
+    int kalmanBallIdx = 0;
+    for (KalmanBall& kalmanBall : kalmanBallList) {
+        std::list<CameraBall>& measurementBalls = appliedBallsList.at(kalmanBallIdx);
+
+        int cameraBallIdx = 0;
+        for (const CameraBall& cameraBall : ballList) {
+            double dist = (kalmanBall.getPos() - cameraBall.getPos()).mag();
+
+            // Increase the distance of our cutoff by the velocity
+            // This is so the ball doesn't move outside the kalman filter position radius when the ball
+            // instantly stops (like in sim)
+            if (dist < *MHKF_radius_cutoff + kalmanBall.getVel().mag()) {
+                measurementBalls.push_back(cameraBall);
+                usedCameraBall.at(cameraBallIdx) = true;
+            }
+            cameraBallIdx++;
+        }
+
+        kalmanBallIdx++;
+    }
+
+    // Apply the ball measurements to the kalman filters
+    kalmanBallIdx = 0;
+    for (KalmanBall& kalmanBall : kalmanBallList) {
+        std::list<CameraBall>& measurementBalls = appliedBallsList.at(kalmanBallIdx);
+
+        // We had at least one measurement near this ball
+        if (measurementBalls.size() > 0) {
+            CameraBall avgBall = CameraBall::CombineBalls(measurementBalls);
+            kalmanBall.predictAndUpdate(calcTime, avgBall);
+
+        // There aren't any measurements so just predict
+        } else {
+            kalmanBall.predict(calcTime);
+        }
+    }
+
+    // Any balls not used, create a kalman ball at that position
```

Comment:
Yep, if there are two measurements of the same physical object, really close together, they would both be used to initialize the filter instead of doing a init then update.